src/libraries/System.Linq/src/System/Linq/MaxMin.cs

```diff
@@ -0,0 +1,118 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Runtime.Intrinsics;
+
+namespace System.Linq
+{
+    public static partial class Enumerable
+    {
+        private interface IMinMaxCalc<T> where T : struct, IBinaryInteger<T>
+        {
+            public static abstract bool Compare(T left, T right);
+            public static abstract Vector128<T> Compare(Vector128<T> left, Vector128<T> right);
+            public static abstract Vector256<T> Compare(Vector256<T> left, Vector256<T> right);
+        }
+
+        private static T MinMaxInteger<T, TMinMax>(this IEnumerable<T> source)
+            where T : struct, IBinaryInteger<T>
+            where TMinMax : IMinMaxCalc<T>
+        {
+            T value;
+
+            if (source.TryGetSpan(out ReadOnlySpan<T> span))
+            {
+                if (span.IsEmpty)
+                {
+                    ThrowHelper.ThrowNoElementsException();
+                }
+
+                if (!Vector128.IsHardwareAccelerated || span.Length < Vector128<T>.Count)
+                {
+                    value = span[0];
+                    for (int i = 1; i < span.Length; i++)
+                    {
+                        if (TMinMax.Compare(span[i], value))
+                        {
+                            value = span[i];
+                        }
+                    }
+                }
+                else if (!Vector256.IsHardwareAccelerated || span.Length < Vector256<T>.Count)
+                {
+                    ref T current = ref MemoryMarshal.GetReference(span);
+                    ref T lastVectorStart = ref Unsafe.Add(ref current, span.Length - Vector128<T>.Count);
+
+                    Vector128<T> best = Vector128.LoadUnsafe(ref current);
+                    current = ref Unsafe.Add(ref current, Vector128<T>.Count);
+
+                    while (Unsafe.IsAddressLessThan(ref current, ref lastVectorStart))
+                    {
+                        best = TMinMax.Compare(best, Vector128.LoadUnsafe(ref current));
+                        current = ref Unsafe.Add(ref current, Vector128<T>.Count);
+                    }
+                    best = TMinMax.Compare(best, Vector128.LoadUnsafe(ref lastVectorStart));
+
+                    value = best[0];
+                    for (int i = 1; i < Vector128<T>.Count; i++)
+                    {
+                        if (TMinMax.Compare(best[i], value))
+                        {
+                            value = best[i];
+                        }
+                    }
```

Comment:
If you'd like to go crazy with vectorization, this reduction could be done like https://gist.github.com/gfoidl/b3da2a1bacbd617c04584b2398cedc6e

* it's branchless*
* has less machine code -- note current JIT unrolls the scalar loops only up to count == 4, then it [emits a loop](https://gist.github.com/gfoidl/b3da2a1bacbd617c04584b2398cedc6e?permalink_comment_id=4314898#gistcomment-4314898)

But the code is ugly, and I don't assume that's the right fit for here. 
But maybe there should be (internal) type with these extensions to vector for reduction (min, max)? Or even a public API?

I don't know how common such use is, in my work reduction for min/max is only used with `double`.

\* except for `long`, `ulong` there's the scalar approach less work to do, as there are only 2 elements in the vector

Reply:
> If you'd like to go crazy with vectorization, this reduction could be done like

Thanks. I think I'll refrain, but neat to see :)